ffi/dotnet-idsdk/IoticsIdLib/src/Identity.cs

```diff
@@ -0,0 +1,244 @@
+    /// <param name="cId">a unique ID in the identity document referring the public key matching this identity</param>
+    /// <returns>the twin identity with the delegation already setup</returns>
+    public TwinIdentity CreateTwinIdentityWithControlDelegation(AgentIdentity creator, string cKeyName, string cId)
+    {
+        string did = Tools.InvokeGoFunction(() => IdLib.CreateTwinDidWithControlDelegation(
+            ResolverAddress, creator.Did, creator.KeyName, creator.Id, creator.Seed, cKeyName, cId
+        ));
+        return new TwinIdentity(this, creator, cKeyName, cId, did);
+    }
+
+    /// <summary>
+    /// Creates a delegation from a user to an agent.
+    /// </summary>
+    /// <param name="agentIdentity">the identity of the agent being delegated</param>
+    /// <param name="userIdentity">the identity of the delegating user</param>
+    /// <param name="cDelegationName">the unique name of this delegation</param>
+    /// <returns>throws DelegationException if an error occurs</returns>
+    internal void UserDelegatesAuthenticationToAgent(
+        AgentIdentity agentIdentity,
+        UserIdentity userIdentity,
+        string cDelegationName
+    ) {
+        string? result = Tools.InvokeGoFunction(() => IdLib.UserDelegatesAuthenticationToAgent(ResolverAddress, 
+            agentIdentity.Did, agentIdentity.KeyName, Tools.EnsureHashPrefix(agentIdentity.Id), agentIdentity.Seed,
+            userIdentity.Did, userIdentity.KeyName, Tools.EnsureHashPrefix(userIdentity.Id), userIdentity.Seed,
+            Tools.EnsureHashPrefix(cDelegationName)));
+        
+        if(result != null) {
+            throw new DelegationException(result);
+        }
+    }
+
+    /// <summary>
+    /// In case another agent (other than the creator) needs delegation. 
+    /// This method is of limited use at this stage since, for it to work, both keys of the principals need to be available.
+    /// It should only be sufficient for the agent to submit a proof which is a public value. 
+    /// 
+    /// </summary>
+    /// <param name="agentIdentity">the identity of the agent being delegated</param>
+    /// <param name="twinIdentity">the identity of the delegating twin</param>
+    /// <param name="cDelegationName">the unique name of this delegation</param>
+    /// <returns>throws DelegationException if an error occurs</returns>
+    internal void TwinDelegatesControlToAgent(
+        AgentIdentity agentIdentity,
+        TwinIdentity twinIdentity,
+        string cDelegationName
+    ) {
+        string? result = Tools.InvokeGoFunction(() => IdLib.TwinDelegatesControlToAgent(ResolverAddress, 
+            agentIdentity.Did, agentIdentity.KeyName, Tools.EnsureHashPrefix(agentIdentity.Id), agentIdentity.Seed,
+            twinIdentity.Did, twinIdentity.KeyName, Tools.EnsureHashPrefix(twinIdentity.Id), twinIdentity.Seed,
+            Tools.EnsureHashPrefix(cDelegationName)));
+        if(result != null) {
```

Comment:
Shouldn't a level above be throwing exceptions and this class only is a pass-through wrapper?

Reply:
I repeatedly endup writing the ensure in the application - i find it counter intuitive to have to prepend the string with a #

I am not strong about it but if you prefer not to do it in the Golang layer then we need to document it

Reply:
> if you prefer not to do it in the Golang layer

Maybe I worded that badly - I am suggesting we **do** validate extra stuff in golang level (not even FFI) so that each of the FFI wrappers can be thinner and validation is consistent across languages.